src/internal_math.rs

```diff
@@ -0,0 +1,224 @@
+        /// NOTE: reconsider after Ice Lake
+        struct Barrett {
+            _m: u32,
+            im: u64,
+        }
+
+        impl Barrett {
+            /// # Arguments
+            /// * `m` `1 <= m`
+            fn new(m: u32) -> Barrett {
+                Barrett {
+                    _m: m,
+                    im: (-1i64 as u64) / (m as u64) + 1,
+                }
+            }
+
+            /// # Returns
+            /// `m`
+            fn umod(&self) -> u32 {
+                self._m
+            }
+
+            /// # Parameters
+            /// * `a` `0 <= a < m`
+            /// * `b` `0 <= b < m`
+            ///
+            /// # Returns
+            /// a * b % m
+            fn mul(&self, a: u32, b: u32) -> u32 {
+                // [1] m = 1
+                // a = b = im = 0, so okay
+
+                // [2] m >= 2
+                // im = ceil(2^64 / m)
+                // -> im * m = 2^64 + r (0 <= r < m)
+                // let z = a*b = c*m + d (0 <= c, d < m)
+                // a*b * im = (c*m + d) * im = c*(im*m) + d*im = c*2^64 + c*r + d*im
+                // c*r + d*im < m * m + m * im < m * m + 2^64 + m <= 2^64 + m * (m + 1) < 2^64 * 2
+                // ((ab * im) >> 64) == c or c + 1
+                let mut z = a as u64;
+                z *= b as u64;
+                let x = (((z as u128) * (self.im as u128)) >> 64) as u64;
+                let mut v = (z - x * self._m as u64) as u32;
+                if self._m <= v {
+                    v += self._m;
+                }
+                v
+            }
+        }
+
+        /// # Parameters
+        /// * `n` `0 <= n`
+        /// * `m` `1 <= m`
+        ///
+        /// # Returns
+        /// `(x ** n) % m`
+        /* const */
+        fn pow_mod_constexpr(x: i64, mut n: i64, m: i32) -> i64 {
+            if m == 1 {
+                return 0;
+            }
+            let _m = m as u32;
+            let mut r: u64 = 1;
+            let mut y: u64 = safe_mod(x, m as i64) as u64;
+            while n != 0 {
+                if (n & 1) > 0 {
+                    r = (r * y) % (_m as u64);
+                }
+                y = (y * y) % (_m as u64);
+                n >>= 1;
+            }
+            r as i64
+        }
+
+        /// Reference:
+        /// M. Forisek and J. Jancina,
+        /// Fast Primality Testing for Integers That Fit into a Machine Word
+        ///
+        /// # Parameters
+        /// * `n` `0 <= n`
+        /* const */
+        fn is_prime_constexpr(n: i32) -> bool {
+            let n = n as i64;
+            match n {
+                _ if n <= 1 => return false,
+                2 | 7 | 61 => return true,
+                _ if n % 2 == 0 => return false,
+                _ => {}
+            }
+            let mut d = n - 1;
+            while d % 2 == 0 {
+                d /= 2;
+            }
+            for a in [2, 7, 61].iter().copied() {
```

Comment:
The following is more concise:
```suggestion
            for &a in &[2, 7, 61] {
```

Reply:
Thanks, I didn't know that expression